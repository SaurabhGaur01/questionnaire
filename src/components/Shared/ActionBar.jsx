import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { setActiveStep } from '../../ducks/activeStep';

const ActionBar = ({
    activeStep,
    handleComplete,
    stepsLength,
    completedSteps,
    actionSetActiveStep,
}) => {
    const handleBack = () => {
        actionSetActiveStep(activeStep - 1);
    };

    return (
        <div className="action-buttons">
            <button 
                id ="back-button" 
                type="button" 
                className="btn btn-secondary"
                disabled={activeStep === 0} 
                onClick={handleBack}
            >
                Back
            </button>
            <button 
                id ="complete-button" 
                type="button" 
                className="btn btn-primary"
                onClick={handleComplete}
            >
            {
                Object.keys(completedSteps).length === stepsLength - 1 
                ? 'Finish' : 'Next'
            }
            </button>
        </div> 
    )
}

ActionBar.propTypes = {
    stepsLength: PropTypes.number.isRequired,
    activeStep: PropTypes.number.isRequired,
    handleComplete: PropTypes.func.isRequired,
    actionSetActiveStep: PropTypes.func.isRequired,
    completedSteps: PropTypes.shape({}).isRequired,
}

const mapDispatchAsProps = {
    actionSetActiveStep: setActiveStep,
}

export const mapStateToProps = state => ({
    stepsLength: state.stepsLength,
    completedSteps: state.completedSteps,
    activeStep: state.activeStep,
});

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { ActionBar as TestableActionBar };

export default hocChain(ActionBar);