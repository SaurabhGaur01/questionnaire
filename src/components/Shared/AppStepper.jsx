import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import Stepper from '@material-ui/core/Stepper';
import Step from '@material-ui/core/Step';
import StepButton from '@material-ui/core/StepButton';
import { setActiveStep } from '../../ducks/activeStep';

const AppStepper = ({ 
    activeStep,
    completedSteps,
    steps,
    actionSetActiveStep
}) =>{
    const handleStep = (step) => () => {
        actionSetActiveStep(step);
    };

    return (
        <Stepper nonLinear activeStep={activeStep}>
            {steps.map((label, index) => (
                <Step key={label}>
                    <StepButton id="step-button" onClick={handleStep(index)} completed={completedSteps[index]}>
                        {label}
                    </StepButton>
                </Step>
            ))}
        </Stepper>
    );
}    

AppStepper.propTypes = {
    activeStep: PropTypes.number.isRequired,
    completedSteps: PropTypes.shape({}).isRequired,
    steps: PropTypes.arrayOf(PropTypes.string).isRequired,
    actionSetActiveStep: PropTypes.func.isRequired,
}

const mapDispatchAsProps = {
    actionSetActiveStep: setActiveStep,
}

export const mapStateToProps = state => ({
    completedSteps: state.completedSteps,
    activeStep: state.activeStep,
});

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { AppStepper as TestableAppStepper };

export default hocChain(AppStepper);