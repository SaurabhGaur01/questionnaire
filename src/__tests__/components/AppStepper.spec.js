import React from 'react';
import { shallow } from 'enzyme';
import { TestableAppStepper, mapStateToProps} from '../../components/Shared/AppStepper';

describe('AppStepper Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            actionSetActiveStep: jest.fn(),
            steps: ['step-1', 'step-2', 'step-3'],
            activeStep: 2,
            completedSteps: {mockkey: 'step-1'},
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableAppStepper {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call actionSetActiveStep when any step button is clicked', () => {
        const renderedModule = shallow(<TestableAppStepper {...props} />);
        renderedModule.find('#step-button').at(0).simulate('click');
        expect(props.actionSetActiveStep).toHaveBeenCalledTimes(1);
    });

    describe('mapStateToProps', () => {
        const completedStepsMock = {
            mockkey: 'mockValue',
        };
        const activeStepMock = 1;

        const mockedState = {
            completedSteps: completedStepsMock,
            activeStep: activeStepMock,
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    completedSteps: completedStepsMock,
                    activeStep: activeStepMock,
                }
            )
        })
    })
})