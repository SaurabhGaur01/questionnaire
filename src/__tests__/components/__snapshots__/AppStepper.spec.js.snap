// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppStepper Component should render correctly 1`] = `
<WithStyles(ForwardRef(Stepper))
  activeStep={2}
  nonLinear={true}
>
  <WithStyles(ForwardRef(Step))
    key="step-1"
  >
    <WithStyles(ForwardRef(StepButton))
      id="step-button"
      onClick={[Function]}
    >
      step-1
    </WithStyles(ForwardRef(StepButton))>
  </WithStyles(ForwardRef(Step))>
  <WithStyles(ForwardRef(Step))
    key="step-2"
  >
    <WithStyles(ForwardRef(StepButton))
      id="step-button"
      onClick={[Function]}
    >
      step-2
    </WithStyles(ForwardRef(StepButton))>
  </WithStyles(ForwardRef(Step))>
  <WithStyles(ForwardRef(Step))
    key="step-3"
  >
    <WithStyles(ForwardRef(StepButton))
      id="step-button"
      onClick={[Function]}
    >
      step-3
    </WithStyles(ForwardRef(StepButton))>
  </WithStyles(ForwardRef(Step))>
</WithStyles(ForwardRef(Stepper))>
`;
